Part I: Conquering our MongoDB Timeout (and Other Optimizations)

galacemiguel opened the pull request:
### ⁉️ About
Just wanted to share what I learned having dealt with this issue! Might be useful to future devs that this be documented. 😄

Outline:
1. [Avoid `map`-ing over a Ruby array and use `FastQuery`'s `values` instead](https://github.com/quipper/quipper.github.io/blob/4608c65a20d45b6386d0cf219d3217f7561a881a/_posts/2019-04-23-mongodb-timeout-i.markdown#avoid-map-ing-over-a-ruby-array-and-use-fastquerys-values-instead)
2. [Consider the Aggregation Pipeline if you need distinct values](https://github.com/quipper/quipper.github.io/blob/4608c65a20d45b6386d0cf219d3217f7561a881a/_posts/2019-04-23-mongodb-timeout-i.markdown#consider-the-aggregation-pipeline-if-you-need-distinct-values)
3. [Bonus: Cache the result of expensive computations](https://github.com/quipper/quipper.github.io/blob/4608c65a20d45b6386d0cf219d3217f7561a881a/_posts/2019-04-23-mongodb-timeout-i.markdown#bonus-cache-the-result-of-expensive-computations)

You may read the post [here](https://github.com/quipper/quipper.github.io/blob/4608c65a20d45b6386d0cf219d3217f7561a881a/_posts/2019-04-23-mongodb-timeout-i.markdown). Comments are much appreciated!

kristm commented:
you should add yourself here https://github.com/quipper/quipper.github.io/blob/master/_config.yml#L64

galacemiguel commented:
Did a major rewrite following @rbmrclo's concerns! Addressed your comment as well, @kristm!

Please check if you have time, @quipper/ph-devs. Thank you! 🙇

rbmrclo commented on _posts/2019-04-23-mongodb-timeout-i.markdown:
```diff
@@ -0,0 +1,153 @@
+
+# If you need distinct values, write your own Plucky query!
+
+While `FastQuery` also has a `distinct` method, unfortunately, it doesn't leverage `Plucky`'s native `distinct` querying capabilities. [Peeking into the source code](https://github.com/quipper/monorepo/blob/develop/schema/lib/schema/fast_query.rb), we can see that `FastQuery#distinct` is really just a wrapper over the `values` method discussed earlier, chained with an `Array#uniq`:
```
This code is not exposed to the internet since it's a private repo. I think it's useless to add it as a reference here 🙇

rbmrclo replied:
What I would recommend is to create a [gist](https://gist.github.com) and extract the code that you only need to demonstrate. For example:

----

"Here's a snippet of how FastQuery#distinct looks like in codebase:"

```ruby
class FastQuery

  attr_reader :plucky

  def initialize(plucky)
    @plucky = plucky
  end

  # ... code omitted

  def distinct
     # code goes here and use .plucky
  end

end
```

rbmrclo replied:
Lol after writing this comment and scrolled down a bit, I realised that there's already a snippet of code.
- https://github.com/quipper/quipper.github.io/pull/35/files#diff-9204946012b90d1c3e7af2bb81d37b55R72

Sorry! 😅

rbmrclo replied:
Anyway, just remove the link as a reference then it's all good 👍

rbmrclo commented on _posts/2019-04-23-mongodb-timeout-i.markdown:
````diff
@@ -0,0 +1,153 @@
+Fortunately, we can get around this by writing our own `Plucky` query, without having to interface with `FastQuery`:
+
+```ruby
+Plucky::Query.new(<SOME_MODEL>.collection)
````
```suggestion
Plucky::Query.new(@model.collection)
```

rbmrclo replied:
Maybe it's just me but the `<SOME_MODEL>` approach looks ugly for me. 😅 

I think it's more intuitive to use `@model` since it's a native ruby code and it's also the kind of approach that you would want to emphasize here. (i.e "An instance of a class (model) with `.collection` method")

galacemiguel replied:
Yes, thank you! I've been wanting to find a better way of writing this! 🤩

galacemiguel commented:
Thanks for taking another look, @rbmrclo! Addressed your comments just now. 🙇

rbmrclo commented:
LGTM 👍 🥇
